### DISCUSSION TITLE
Mitsuba 3 Spectral Rendering: setting the number of spectral samples

### DISCUSSION THREAD

**Discussion by juanraul8** [mitsuba-renderer/mitsuba3#361] [Q&A]
Dear Mitsuba 3 team,

Is there a way to define the number of spectral samples which are going to be used for the render? Similar to the setting parameter SPECTRUM_SAMPLES in Mitsuba 0.6. 

I tried to compile the _scalar_spectral_ with the following configuration:

```
 "scalar_spectral": {
        "float": "float",
        "spectrum": "Color<Float, 10>"
    },
```
Unfortunately, I got the following error:

```
ImportError: pybind11::detail::get_type_info: unable to find type info for "drjit::Array<float, 10ul>"
```

Best regards,
Juan Raul Padron Griffe

**Comment by njroussel** [comment]
Hi @juanraul8

In my limited experience with spectral rendering and Mitsuba 0.6, my understanding is the following:

In Mitsuba 3 (and Mitsuba 2), the wavelengths are defined at runtime and are just an additional dimension to sample and integrate over. Specifically, during the ray generation we make use of [Endpoint.sample_ray()](https://mitsuba.readthedocs.io/en/stable/src/api_reference.html#mitsuba.Endpoint.sample_ray) which has as its second argument a "_uniformly distributed 1D value that is used to sample the spectral dimension of the emission profile._". In fact, usually this means that its second argument is used to make a call to [Endpoint.sample_wavelengths()](https://mitsuba.readthedocs.io/en/stable/src/api_reference.html#mitsuba.Endpoint.sample_wavelengths).

Now for your compilation issue, I might have misunderstood what you did, but you should just have to add `"scalar_spectral"` to the `"enabled"` list of variants if it is not already in it.
The snippet you posted is the template definition that will then be "injected" into Mitsuba. You can learn a bit more about this [here](https://mitsuba.readthedocs.io/en/stable/src/developer_guide/variants_cpp.html).

For `spectral` variants, you'll see that the template definition is usually something like this:
```
    "*_spectral": {
        "float": ...,
        "spectrum": "Spectrum<Float, 4>"
    },
```
The number 4 defines the number of wavelengths per ray, and it is somewhat of an arbitrary choice. You should be able to increase/decrease the number depending on whether or not a single ray (and a single path) should carry more/less spectral information in your scene/setup.

**Comment by juanraul8** [comment]
Hi @njroussel

Thanks for the explanation!. I actually wrote the code snippet wrong in the message above. The idea was tracking more than 4 wavelengths, lets say 10. Below you can find the mitsuba.conf file I tried, in case you want to reproduce the issue.

{
   ...

    "enabled": [ "scalar_rgb", "scalar_spectral" ],

    # The variant below will automatically be set when importing the Python
    # mitsuba module. If the value is "", the desired variant will need to be
    # specified using "mitsuba.set_variant(..)" before importing any packages.

    "python-default": "",

    # -------------------------------------------------------------
    # The remainder of this file defines the variants and does not
    # need to be edited (unless you wish to create new variants).
    # -------------------------------------------------------------

    # Scalar variant definitions
    ....

    "scalar_spectral": {
        "float": "float",
        "spectrum": "Spectrum<Float, 10>"
    },

    "scalar_spectral_double": {
        "float": "double",
        "spectrum": "Spectrum<Float, 10>"
    },

    "scalar_spectral_polarized": {
        "float": "float",
        "spectrum": "MuellerMatrix<Spectrum<Float, 10>>"
    },

    "scalar_spectral_polarized_double": {
        "float": "double",
        "spectrum": "MuellerMatrix<Spectrum<Float, 10>>"
    },

    # LLVM variant definitions
    ...
}

If I tried to compile Mitsuba 3 with this configuration, I got the following error:

`[ImportError: pybind11::detail::get_type_info: unable to find type info for "drjit::Array<float, 10ul>"](url)`

If I write 3 o 4 wavelengths, the code compiles fine!

**Comment by Speierers** [comment]
This is probably because Dr.Jit only implements Python bindings for a limited set of `drjit::Array` (e.g. `Array<float/int, 3/4>`). You will probably need to add those bindings to get this to work.

Take a look at `drjit/src/python/bind.h`, it defines the macro `DRJIT_BIND_ARRAY_TYPES` that generates bindings for the arrays. You could add the following line to add binding for the `drjit::Array<float, 10>` type:
```c++
    DRJIT_BIND_ARRAY_TYPES_DIM(Module, Guide, Scalar, 10)
```

Although this might not be enough.
